pkg/plugins/placementpolicy/core/policy.go

```diff
@@ -0,0 +1,233 @@
+package core
+
+import (
+	"sync"
+
+	"github.com/Azure/placement-policy-scheduler-plugins/apis/v1alpha1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/kubernetes/pkg/scheduler/framework"
+)
+
+type PolicyInfo struct {
+	//policy's namespace - from CRD
+	Namespace     string
+	//policy's name - from CRD
+	Name          string
+	//policy's action - from CRD
+	Action        v1alpha1.Action
+	//policy's target - from CRD
+	TargetSize    *intstr.IntOrString
+	//collection of pods matched if the policy is `BestEffort`; **not** used for computation
+	matchedPods   sets.String
+	//collection of pods that could be managed by the policy; used for computation as the **total**
+	qualifiedPods sets.String
+	//collection of pods assigned to a node according to the policy; used for computation
+	managedPods   sets.String
+	//does the ratio of managed-to-qualified meet the `TargetSize`
+	targetMet     bool
+}
+
+func newPolicyInfo(namespace string, name string, action v1alpha1.Action, targetSize *intstr.IntOrString) *PolicyInfo {
+	policy := &PolicyInfo{
+		Namespace:     namespace,
+		Name:          name,
+		Action:        action,
+		TargetSize:    targetSize,
+		matchedPods:   sets.NewString(),
+		qualifiedPods: sets.NewString(),
+		managedPods:   sets.NewString(),
+		targetMet:     false,
+	}
+	return policy
+}
+
+type PolicyInfos struct {
+	sync.RWMutex
+	//map (by `Namespace`) of map (by `Name`) of `PolicyInfo`
+	internal map[string]map[string]*PolicyInfo
+}
+
+func NewPolicyInfos() PolicyInfos {
+	return PolicyInfos{
+		internal: make(map[string]map[string]*PolicyInfo),
+	}
+}
+
+// GetPolicy gets the PolicyInfo if it exists, otherwise it creates and returns the newly created PolicyInfo
+func (pi *PolicyInfos) GetPolicy(policyNamespace string, policyName string, action v1alpha1.Action, targetSize *intstr.IntOrString) *PolicyInfo {
+	var policy *PolicyInfo = nil
+	var policyExists bool = false
+
+	pi.RLock()
+	namespacePolicies, namespaceExists := pi.internal[policyNamespace]
+	if namespaceExists {
+		policy, policyExists = namespacePolicies[policyName]
+	}
+	pi.RUnlock()
+
+	if policyExists {
+		return policy
+	}
+
+	pi.Lock()
+	if !namespaceExists {
+		pi.internal[policyNamespace] = make(map[string]*PolicyInfo)
+	}
+
+	created := newPolicyInfo(policyNamespace, policyName, action, targetSize)
+	pi.internal[policyNamespace][policyName] = created
+
+	pi.Unlock()
+
+	return created
+}
+
+// RemovePolicy removes the policy from the in-memory collection
+func (pi *PolicyInfos) RemovePolicy(policyNamespace string, policyName string) {
+	pi.Lock()
+	delete(pi.internal[policyNamespace], policyName)
```

Comment:
Should we delete namespace from the map if the name map for it is empty after delete? This will ensure the cache doesn't go indefinitely when policies in namespace are deleted.

Reply:
Agreed